### PR TITLE
Datacube resample

### PR DESCRIPTION
**Pull request by rcooke-ast**
This (long overdue) PR includes support for new algorithms for the coadd3d module. I've included support for resampling the 3D voxels (very slow), and a faster subsampling approach, which is now the default option. This will need to survive the dev suite, and I'll run that once folks are happy with these changes.

I've also removed the skimage dependency from one part of the code.

### CODE REVIEW

File: pypeit/alignframe.py
Comment (kbwestfall) [inline pypeit/alignframe.py]: Put these lines in a code block?  And give the shape of the traces array.

File: pypeit/alignframe.py
Comment (kbwestfall) [inline pypeit/alignframe.py]: Doc incomplete.

File: pypeit/core/extract.py
Comment (kbwestfall) [inline pypeit/core/extract.py:691]: Good catch!

File: pypeit/par/pypeitpar.py
Comment (kbwestfall) [inline pypeit/par/pypeitpar.py]: "... voxels, the ... " -> "... voxels, but the ..." ?

File: pypeit/slittrace.py
Comment (kbwestfall) [inline pypeit/slittrace.py:469]: Nice!  Much cleaner.

File: pypeit/core/datacube.py
Comment (kbwestfall) [inline pypeit/core/datacube.py]: Put this in a code block.

File: pypeit/core/datacube.py
Comment (kbwestfall) [inline pypeit/core/datacube.py]: Wrong syntax for a Numpy style docstring.  Return entries should be identical to the parameter entries:
```
object_instance_name : type, etc
    description
```

File: pypeit/core/datacube.py
Comment (kbwestfall) [inline pypeit/core/datacube.py]: Is this function necessary?  All the input checks should also be checked when instantiating the `AlignmentSplines` object, and I think you only call this function once.  Can you instead instantiate the `AlignmentSplines` object directly at the appropriate spot?

File: pypeit/core/datacube.py
Comment (kbwestfall) [inline pypeit/core/datacube.py]: Why is the default name of the spectrograph `PYP_SPEC`?  Does the code fault if you set it to None?

File: pypeit/core/datacube.py
Comment (kbwestfall) [inline pypeit/core/datacube.py]: "sli" -> "slit"

File: pypeit/core/datacube.py
Comment (kbwestfall) [inline pypeit/core/datacube.py]: This one and the previous argument is optional.

File: pypeit/core/datacube.py
Comment (kbwestfall) [inline pypeit/core/datacube.py]: What are the units of `fluximg` if `fluxcal` is false?

File: pypeit/core/datacube.py
Comment (kbwestfall) [inline pypeit/core/datacube.py]: Doesn't this mean the code always faults when using this function?  If so, I think we shouldn't allow users to select this method.

File: pypeit/core/datacube.py
Comment (kbwestfall) [inline pypeit/core/datacube.py]: These need to be imported at the top of the file, and we need a check early in any script workflow that checks that `shapely` is installed if a user selects the resample method.

File: pypeit/core/datacube.py
Comment (kbwestfall) [inline pypeit/core/datacube.py]: Don't need the else (`msgs.error` breaks the code flow).

File: pypeit/core/datacube.py
Comment (kbwestfall) [inline pypeit/core/datacube.py]: Given the overlap with the new subsample function, do we still need this ngp function?

**Review by kbwestfall — CHANGES_REQUESTED** [review]
Thanks @rcooke-ast .  Mostly minor stuff from me and some clean-up requests.  Two more significant requests are:

1. Tests: We need some unit tests and/or vet tests for the resampling functionality.  Is there an associated dev-suite PR incoming?
2. Docs: It would be really helpful to have a complete explanation of how the datacube reconstruction is performed.  I'm not sure how much of this already exists, but it would be good to have an explanation in the main doc pages describing the different algorithms available and their pros/cons.

File: pypeit/par/pypeitpar.py
Comment (MNBrod) [inline pypeit/par/pypeitpar.py]: Is there really a reason to leave resampling in as an option if the result of subsampling is essentially identical and faster? If so, documenting that reason (e.g. "this weird edge case" or "if you really care about ___") would be welcome

File: pypeit/alignframe.py
Comment (rcooke-ast) [reply pypeit/alignframe.py]: Great - done!  But I may not have done this right - please let me know if anything is out of place  :-)

File: pypeit/alignframe.py
Comment (rcooke-ast) [reply pypeit/alignframe.py]: Whoops - thanks for spotting that!

File: pypeit/par/pypeitpar.py
Comment (rcooke-ast) [reply pypeit/par/pypeitpar.py]: done

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Nope... it's pointless... I'll remove it - thanks!!

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Fixed - Changed to None. The code works just fine with `PYP_SPEC=None`

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Added some docs for this

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: It does... I remember now that this is a work in progress, that took a lot of effort to get working and there's a bug I can't resolve. For now, I'll comment out the `resample` option, but I'd like to keep the code here for the time-being.

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: That's right... I was still debugging this code, but for now, I've commented it all out.

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Resample is now gone as an option!

File: pypeit/par/pypeitpar.py
Comment (rcooke-ast) [reply pypeit/par/pypeitpar.py]: Good point, thanks!  I've also realised that the resampling code has a bug that I haven't managed to track down, so I'm removing this support for now, anyway. I don't really think there is an edge case, tbh...

**Comment by rcooke-ast** [comment]
Thanks @kbwestfall and @MNBrod for the quick reviews!

> Tests: We need some unit tests and/or vet tests for the resampling functionality. Is there an associated dev-suite PR incoming?

There is already a test in the dev-suite for the subsample algorithm, and I've now removed the resample algorithm.

> Docs: It would be really helpful to have a complete explanation of how the datacube reconstruction is performed. I'm not sure how much of this already exists, but it would be good to have an explanation in the main doc pages describing the different algorithms available and their pros/cons.

Agreed... I'll write that now and push the changes.

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Yes... for the moment... This function is used when combining multiple cubes together. The subsample function is only used for generating a NGP cube when converting single spec2d files into a single datacube. I can't find a resource friendly way of doing this to multiple frames, yet.

**Review by rcooke-ast — COMMENTED** [review]
Thanks for the comments!

File: pypeit/alignframe.py
Comment (kbwestfall) [reply pypeit/alignframe.py]: The syntax looks right, but you might need some empty lines above and below the code in the block.  I.e., something like:
```
    arg (type):
        Here's the description, and some code:
        
        .. code-block:: python

            this = 'the code'

    next_arg (type):
        ...
```
I would compile the docs, and check what the result looks like, assuming you don't get any warnings.

**Review by kbwestfall — APPROVED** [review]
Thanks for addressing all my comments!  Have you run any of the tests?

File: pypeit/alignframe.py
Comment (rcooke-ast) [reply pypeit/alignframe.py]: OK - it compiled the docs just fine (but I decided to add the space anyway, since it looks a lot clearer like this)

**Comment by rcooke-ast** [comment]
> Thanks for addressing all my comments! Have you run any of the tests?

Not yet - I figured I would wait for a review from @jhennawi, implement all of the suggested changes, and then fix any problems.

File: pypeit/core/datacube.py
Comment (jhennawi) [inline pypeit/core/datacube.py]: Remove?

File: pypeit/core/datacube.py
Comment (jhennawi) [inline pypeit/core/datacube.py]: Move to deprecated?

File: pypeit/core/datacube.py
Comment (jhennawi) [inline pypeit/core/datacube.py]: We use ivar instead of var everywhere in PypeIt for the reason that if ivar = 0 that is better than var = 0 to describe bad pixels. Of course we also have masks that deal with that. But my point is that you are switching convention here by outputting a variance instead of an inverse variance.

File: pypeit/core/datacube.py
Comment (jhennawi) [inline pypeit/core/datacube.py]: Maybe add to the docs that the if subsample > 10 the errors will be correlated and the covariance is not being tracked, so the errors won't be right. That is fine, tracking the covariance would be massively painful and this is the tradeoff one has to make between sampling and better looking cubes, and poor sampling and better behaved errors. I would just add a note. Finally, in coadd2d, we called subsample spec_fact and spat_fact. And I think that definition is the inverse of what you have here. It might be best to try to adopt consistent definitions. In other words, spat_fact = 2.0 performs a coarse rebidding, whereas spat_fact = 0.5 would be your subsample = 2. Finally, shouldn't there be an option here to have a different sampling in the spectral direction (is that your z) as opposed to the spatial directions (x,y)?

File: pypeit/core/datacube.py
Comment (jhennawi) [inline pypeit/core/datacube.py]: See my comment above. Would it not be best to distinguish between spectral directions and spatial directions in your documentation to make things more clear.

File: pypeit/par/pypeitpar.py
Comment (jhennawi) [inline pypeit/par/pypeitpar.py:1398]: Nice docs. I suggest adding what you wrote here to the core routine to address my comment above.

Comment (jhennawi) [reply pypeit/par/pypeitpar.py:1398]: Maybe clarify that when multiple pixels hit the same bin for NGP, you are taking the average flux. I think that SB and hence flux are still conserved, but maybe worth a note since your docs suggest it would just add the flux.

**Review by jhennawi — CHANGES_REQUESTED** [review]
Looks great @rcooke-ast. I have some rather minor comments about definitions and docs. Otherwise you are good to go.

File: pypeit/par/pypeitpar.py
Comment (rcooke-ast) [reply pypeit/par/pypeitpar.py:1398]: Yes, good catch... I'll add that in.

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Removed

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Yes, good idea - moved (and I've moved the shapely import, too).

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: OK - I'm willing to concede that `varcube` is a poor choice. However, the default thing that we output in the spec1d files include things like `OPT_FLAM`, `OPT_FLAM_IVAR`, and `OPT_FLAM_SIG`. I'm more than happy to output the IVAR and the SIG cubes, but these are usually quite big files. Also, I suspect `SIG` might be more useful to users than `IVAR` with the caveat that you mention about 0 values in the `SIG` cubes. I'm also not outputting a mask for the datacubes... Maybe we can discuss this on Slack.

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Good point - I've added an important note to the top of the doc string.

Regarding the sampling factor, I can see the benefit of it being consistent, however, it really should be an integer in this case (i.e. you want to break up a pixel into an integer number of small pixels. Also, in datacube, subsample cannot be less than 1. So it needs to be an integer >= 1). We could consider refactoring coadd2d to account for this, but I feel this might be more effort than it's worth... Happy to hear your thoughts on this. One possibility is to adopt the same formalism as in `coadd2d`, but convert this to an (allowed) integer within datacube. Seems like a messy way of doing it, though...

Good idea to have separate x and y subsampling. Note that the subsampling done here is on each individual detector pixel, not quite the spatial and wavelength coordinates (due to the tilts). In my latest refactoring, I've adopted the names `spec_subsample` and `spat_subsample`.

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: I'm not sure I understand this comment in the context of what is written on this line. I've tried to clarify this part of the docstring, but if I've missed the point, let me know!

**Review by rcooke-ast — COMMENTED** [review]
Addressed comments from JFH

File: pypeit/core/datacube.py
Comment (jhennawi) [reply pypeit/core/datacube.py]: Okay, I see your point about this being different since it has to be an integer. In that case, casting an integer to 1/integer is a bad idea, so let's leave your definitions and variables as they are. So shouldn't this be called supersample though? You are not subsampling if your final cube has higher resolution than your individual frames, you are super-sampling right?

**Review by jhennawi — APPROVED** [review]
@rcooke-ast I'm approving but please see my last comment. Also as per the Slack, I'm still for including the ivar as an output, but we can discuss there.

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: So, the final cube doesn't have higher spatial resolution. I've just divided each detector pixel into N subpixels (say 25), and a fraction of those 25 subpixels are put into a single voxel. For example, of the 25 subpixels, 2 subpixels end up in voxelA, 16 subpixels end up in voxelB, and the remaining 7 subpixels end up in voxelC... with NGP, all 25 subpixels (i.e. the 1 pixel) end up in voxelB. So, this is a way to share the flux from 1 detector pixel into several voxels, depending on where the voxel boundaries are.

... and you're right - subsample is the incorrect word... I've now opted for subpixel.

**Comment by rcooke-ast** [comment]
> @rcooke-ast I'm approving but please see my last comment. Also as per the Slack, I'm still for including the ivar as an output, but we can discuss there.

As an alternative, instead of including `ivar` as an output, can we instead include it as a @property. Here is an example:

image

So, the user won't know anything different (unless they go digging into the fits files), and it saves on output filesize.

**Comment by rcooke-ast** [comment]
Merging - and will reconsider adding ivar in a future PR

**Pull request merged**
